base/abstractarraymath.jl

```diff
@@ -330,89 +308,171 @@ julia> repeat([1 2; 3 4], inner=(2, 1), outer=(1, 3))
-    _rshps((shp..., o[1]), (shp_i..., 1), (), (), tail(o))
-@inline _rshps(shp, shp_i, ::Tuple{}, i, ::Tuple{}) = (n = i[1];
-    _rshps((shp..., n), (shp_i..., n), (), tail(i), ()))
-@inline _rshps(shp, shp_i, ::Tuple{}, i, o) = (n = i[1];
-    _rshps((shp..., n * o[1]), (shp_i..., n), (), tail(i), tail(o)))
-@inline _rshps(shp, shp_i, sz, i, o) = (n = sz[1] * i[1];
-    _rshps((shp..., n * o[1]), (shp_i..., n), tail(sz), tail(i), tail(o)))
-_rshps(shp, shp_i, sz, ::Tuple{}, ::Tuple{}) =
-    (n = length(shp); N = n + length(sz); _reperr("inner", n, N))
-_rshps(shp, shp_i, sz, ::Tuple{}, o) =
-    (n = length(shp); N = n + length(sz); _reperr("inner", n, N))
-_rshps(shp, shp_i, sz, i, ::Tuple{}) =
-    (n = length(shp); N = n + length(sz); _reperr("outer", n, N))
-_reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
-    "($n) cannot be less than number of dimensions of input ($N)"))
-
-_negreperr(n) = throw(ArgumentError("number of $n repetitions" *
-    "cannot be negative"))
-
-@noinline function _repeat(A::AbstractArray, inner, outer)
-    any(<(0), inner) && _negreperr("inner")
-    any(<(0), outer) && _negreperr("outer")
-
-    shape, inner_shape = rep_shapes(A, inner, outer)
-
-    R = similar(A, shape)
-    if any(iszero, shape)
-        return R
+module _RepeatInnerOuter
+
+function repeat(arr; inner=nothing, outer=nothing)
+    check(arr, inner, outer)
+    arr, inner, outer = resolve(arr, inner, outer)
+    repeat_inner_outer(arr, inner, outer)
+end
+
+to_tuple(t::Tuple) = t
+to_tuple(x::Integer) = (x,)
+to_tuple(itr) = tuple(itr...)
+
+filltuple(val, N) = ntuple(_->val, N)
+
+function pad(a::Tuple, b::Tuple, c::Tuple)
+    a1,b1 = pad(a,b)
+    b2,c2 = pad(b1, c)
+    a2,c3  = pad(a1, c2)
+    return a2, b2, c2
+end
+function pad(a::Tuple{}, b::NTuple{N, Any}) where {N}
+    filltuple(1, Val(N)), b
+end
+function pad(a::NTuple{N, Any}, b::Tuple{}) where {N}
+    (a, filltuple(1, Val(N)))
+end
+function pad(a::NTuple{N,Any}, b::NTuple{N,Any}) where {N}
+    (a,b)
+end
+function pad(a::Tuple{}, b::Tuple{})
+    (a,b)
+end
+function pad(a::Tuple, b::Tuple)
+    a1 = first(a)
+    b1 = first(b)
+    a_tail, b_tail = pad(Base.tail(a), Base.tail(b))
+    tuple(a1, a_tail...), tuple(b1, b_tail...)
+end
+
+function resolve(arr::AbstractArray{<:Any, N}, inner::NTuple{N, Any}, outer::NTuple{N,Any}) where {N}
+    arr, inner, outer
+end
+function resolve(arr, inner, outer)
+    dims, inner, outer = pad(size(arr), to_tuple(inner), to_tuple(outer))
+    reshape(arr, dims), inner, outer
+end
+function resolve(arr, inner::Nothing, outer::Nothing)
+    return arr, inner, outer
+end
+function resolve(arr, inner::Nothing, outer)
+    dims, outer = pad(size(arr), to_tuple(outer))
+    reshape(arr, dims), inner, outer
+end
+function resolve(arr, inner, outer::Nothing)
+    dims, inner = pad(size(arr), to_tuple(inner))
+    reshape(arr, dims), inner, outer
+end
+
+function check(arr, inner, outer)
+    if inner != nothing
```

Comment:
It's much faster and easier on the compiler to use `===` when comparing to `nothing`.

```suggestion
    if inner !== nothing
```